### PR TITLE
Adds fix so that restoration does not completely fail if Materialized Views in the backup cannot be refreshed. Closes #1933

### LINKED ISSUE #1933: `steampipe` should not fail restoration if materialized views cannot be refreshed during restoration

### PR DESCRIPTION
**Pull request by binaek**
Depends on #1928

### CODE REVIEW

File: cmd/service.go [hunk 1]
```diff
@@ -405,6 +406,7 @@ func runServiceStatusCmd(cmd *cobra.Command, args []string) {
 	defer func() {
 		utils.LogTime("runServiceStatusCmd end")
 		if r := recover(); r != nil {
+			debug.PrintStack()
```
Comment (kaidaguerre) [inline cmd/service.go:409]: remove

File: cmd/service.go [hunk 2]
```diff
@@ -559,7 +561,9 @@ func showAllStatus(ctx context.Context) {
 func getServiceProcessDetails(process *psutils.Process) (string, string, string, db_local.StartListenType) {
 	cmdLine, _ := process.CmdlineSlice()
 
-	installDir := strings.TrimSuffix(cmdLine[0], db_local.ServiceExecutableRelativeLocation)
+	fmt.Println(db_local.ServiceExecutableRelativeLocation())
```
Comment (kaidaguerre) [inline cmd/service.go:564]: unwanted debug?

File: db/db_local/backup.go [hunk 3]
```diff
@@ -1,20 +1,32 @@
...
+	"github.com/turbot/steampipe/utils"
 )
 
-type dbStartConfig struct {
+var (
+	errDbInstanceRunning = fmt.Errorf("cannot start DB backup - an instance is still running. To stop running services, use %s ", constants.Bold("steampipe service stop"))
+)
+
+const (
+	NoMatViewRefreshListFileName   = "without_refresh.lst"
```
Comment (kaidaguerre) [inline db/db_local/backup.go:25]: do we need to export?

File: db/db_local/backup.go [hunk 4]
```diff
@@ -40,14 +56,31 @@ func prepareBackup(ctx context.Context) (*string, error) {
...
+	return &config.dbName, nil
 }
 
-func takeBackup(ctx context.Context, config *dbStartConfig) error {
+func errIfInstanceRunning(ctx context.Context, location string) error {
+	processes, err := FindAllSteampipePostgresInstances(ctx)
+	if err != nil {
+		return err
+	}
+	for _, p := range processes {
```
Comment (kaidaguerre) [inline db/db_local/backup.go:67]: short comment - what are we doing here?

File: db/db_local/backup.go [hunk 5]
```diff
@@ -79,7 +112,7 @@ func takeBackup(ctx context.Context, config *dbStartConfig) error {
 }
 
 // startDatabaseInLocation starts up the postgres binary in a specific installation directory
-func startDatabaseInLocation(ctx context.Context, location string) (*dbStartConfig, error) {
+func startDatabaseInLocation(ctx context.Context, location string) (*oldDbRunningConfig, error) {
```
Comment (kaidaguerre) [inline db/db_local/backup.go:115]: I meant to query thgis in the other PR - why is dbStartConfig renamed to oldDbRunningConfig?

File: db/db_local/backup.go [hunk 6]
```diff
@@ -129,6 +162,7 @@ func stopDbByCmd(ctx context.Context, cmd *exec.Cmd) error {
 // needsBackup checks whether the `$STEAMPIPE_INSTALL_DIR/db` directory contains any database installation
```
Comment (kaidaguerre) [inline db/db_local/backup.go:162]: remove angle quotes

File: db/db_local/backup.go [hunk 7]
```diff
@@ -160,26 +194,72 @@ func needsBackup(ctx context.Context) (bool, string, error) {
...
 	if err != nil {
 		return err
 	}
-	if info == nil {
+	if runningInfo == nil {
 		return fmt.Errorf("steampipe service is not running")
 	}
 
+	// get the location of the other instance which was backed up
+	_, location, err := needsBackup(ctx)
```
Comment (kaidaguerre) [inline db/db_local/backup.go:214]:
```
_, location, err := needsBackup(ctx)
```
that is a pretty counterintuitive function call

File: db/db_local/backup.go [hunk 8]
```diff
@@ -160,26 +194,72 @@ func needsBackup(ctx context.Context) (bool, string, error) {
...
+		return err
+	}
+
+	// extract the Table of Contents from the Backup Archive
+	toc, err := getTOCFromBackup(ctx)
+	if err != nil {
+		return err
+	}
+
+	// partition the Table of Contents into separate lists (no refresh and only refresh)
```
Comment (kaidaguerre) [inline db/db_local/backup.go:225]: make this comment longer - what is no refresh? why do we want it?

state the problem with mat views and the solution

File: db/db_local/backup.go [hunk 9]
```diff
@@ -160,26 +194,72 @@ func needsBackup(ctx context.Context) (bool, string, error) {
...
+	}
+	err = runRestoreUsingList(ctx, runningInfo, onlyRefreshListFile)
+	if err != nil {
+		// we could not refresh the Materialized views
+		// this is probably because the Materialized views
+		// contain transitive references to unqualified table names
+		// WARN the user.
+		// TODO: We should try to refresh these separately using a Maintenance Client
+		// https://github.com/turbot/steampipe/issues/1934
+		log.Println("[WARN] Could not REFRESH MATERAILIZED VIEWS while restoring data.")
```
Comment (kaidaguerre) [inline db/db_local/backup.go:246]: show error as well?

File: db/db_local/backup.go [hunk 10]
```diff
@@ -202,5 +284,65 @@ func loadBackup(ctx context.Context) error {
 		return err
 	}
 
-	return os.Remove(getBackupLocation())
+	return nil
+}
+
+// splitTocMatViewRefreshData writes back the TableOfContents into a two temporary files:
+// 	1. without REFRESH MATERIALIZED VIEWS
```
Comment (kaidaguerre) [inline db/db_local/backup.go:291]: formatting

File: db/db_local/backup.go [hunk 11]
```diff
@@ -202,5 +284,65 @@ func loadBackup(ctx context.Context) error {
...
+//  2. only REFRESH MATERIALIZED VIEWS
+//
+// This needs to be done because REFRESHING cannot work in restore, since the BackUP will always set
+// a blank search path before commencing and the MATERIALIZED VIEWS may have functions with
+// unqualified table names
+func splitTocMatViewRefreshData(ctx context.Context, toc []string) (string, string, error) {
```
Comment (kaidaguerre) [inline db/db_local/backup.go:297]: awkwards function name - can we improve

Comment (kaidaguerre) [reply db/db_local/backup.go:297]: also inconsistent casing with `getTOCFromBackup(`

File: db/db_local/backup.go [hunk 12]
```diff
@@ -202,5 +284,65 @@ func loadBackup(ctx context.Context) error {
...
+		ctx,
+		pgRestoreBinaryExecutablePath(),
+		databaseBackupFilePath(),
+		// as a tar format
+		"--format=tar",
+		// only the public schema is backed up
+		"--schema=public",
+		"--list",
+	)
+	log.Println("[TRACE]", cmd.String())
```
Comment (kaidaguerre) [inline db/db_local/backup.go:326]: more context for the tracing - or remove it
```
log.Printf("[TRACE] getTOCFromBackup running command %s\n", cmd.String())

File: db/db_local/backup.go [hunk 13]
```diff
@@ -202,5 +284,65 @@ func loadBackup(ctx context.Context) error {
...
+
+	b, err := cmd.Output()
+	if err != nil {
+		return nil, err
+	}
+
+	scanner := bufio.NewScanner(strings.NewReader(string(b)))
+	scanner.Split(bufio.ScanLines)
+
+	lines := []string{";" /* start with an extra comment line */}
```
Comment (kaidaguerre) [inline db/db_local/backup.go:336]: put comment above

File: db/db_local/locations.go [hunk 14]
```diff
@@ -48,9 +50,9 @@ func getDataLocation() string {
 	return loc
 }
 
-// location where the dump file will be stored, so that it can be later restored after connections
+// tar file where the dump file will be stored, so that it can be later restored after connections
```
Comment (kaidaguerre) [inline db/db_local/locations.go:53]: please can we update the other inconsistently named functions in this file

Comment (binaek) [reply db/db_local/locations.go:53]: other PR

File: cmd/service.go (hunk 1 above)
Comment (binaek) [reply cmd/service.go]: doen

File: db/db_local/locations.go (hunk 14 above)
Comment (binaek) [reply db/db_local/locations.go]: https://github.com/turbot/steampipe/issues/1936

File: db/db_local/backup.go (hunk 13 above)
Comment (binaek) [reply db/db_local/backup.go]: done

File: db/db_local/backup.go (hunk 11 above)
Comment (binaek) [reply db/db_local/backup.go]: renamed both these functins

File: db/db_local/backup.go (hunk 3 above)
Comment (binaek) [reply db/db_local/backup.go]: no. done

File: db/db_local/backup.go (hunk 4 above)
Comment (binaek) [reply db/db_local/backup.go]: done

File: db/db_local/backup.go (hunk 5 above)
Comment (binaek) [reply db/db_local/backup.go]: done in the other one

File: db/db_local/backup.go (hunk 6 above)
Comment (binaek) [reply db/db_local/backup.go]: done

File: db/db_local/backup.go (hunk 7 above)
Comment (binaek) [reply db/db_local/backup.go]: changed the fn name

File: db/db_local/backup.go (hunk 9 above)
Comment (binaek) [reply db/db_local/backup.go]: the error can get very long.

File: db/db_local/backup.go (hunk 10 above)
Comment (binaek) [reply db/db_local/backup.go]: done

File: db/db_local/backup.go (hunk 8 above)
Comment (binaek) [reply db/db_local/backup.go:233]: done

Comment (kaidaguerre) [reply db/db_local/backup.go:237]: no it's not?? Still says `// partition the Table of Contents into separate lists (no refresh and only refresh)`

File: db/db_local/backup.go [hunk 15]
```diff
@@ -190,22 +212,92 @@ func needsBackup(ctx context.Context) (bool, string, error) {
...
+	err = runRestoreUsingList(ctx, runningInfo, matviewRefreshListFile)
+	if err != nil {
+		//
+		// we could not refresh the Materialized views
+		// this is probably because the Materialized views
+		// contain transitive references to unqualified table names
+		//
+		// WARN the user.
+		//
+		// TODO: We should try to refresh these separately using a Maintenance Client
```
Comment (kaidaguerre) [inline db/db_local/backup.go:274]: remove this todo

File: db/db_local/backup.go (hunk 12 above)
Comment (kaidaguerre) [reply db/db_local/backup.go:367]: still no response???

**Pull request merged**